### PR TITLE
Drop support for Ember < 3.16 (and upgrade to Ember 3.26)

### PR DESCRIPTION
**Pull request by bendemboski**
Opening for discussion (and to let CI tell me if all the ember-try scenarios pass!)

I propose that we update to the latest and greatest Ember, drop support for Ember 2.18, and do a major version bump. This will give us more flexibility to address embroider compatibility issues, such as #425 and (over time) shed a whole bunch of technical debt.

I plan to follow this PR up with one that adds the embroider compatibility ember-try scenarios.

Thoughts?

### CODE REVIEW

File: README.md
Comment (gilest) [inline README.md:12]: Should we update the supported Ember CLI version also?

**Comment by gilest** [comment]
Thanks for putting this together 👏 

This fulfils a few of the requirements from the v4 roadmap (#331)

Perhaps we should start a `v4` branch where can start landing these PRs.

Would be nice to include Glimmer components and API improvements in a major release since we're making breaking changes.

**Comment by bendemboski** [comment]
@gilest oh, a document roadmap, that seems like a good idea and maybe something I should have read 😆 

Agreed on the glimmer components -- I may have some time to do that soon, but I think it can be a separate PR (and I think a v4 branch makes a lot of sense).

File: README.md
Comment (bendemboski) [reply README.md:12]: That's actually straight out of the [latest blueprint](https://github.com/ember-cli/ember-addon-output/blob/v3.26.0/README.md). I guess very modern addons can still run on a super-old `ember-cli` 🤷

**Comment by bendemboski** [comment]
@gilest question about CodeClimate -- it looks like it's failing because it doesn't like our `.template-lintrc.js`, I'm guessing because its `ember-template-lint` implementation/integration is out-of-date? Now that template linting is part of the CI workflow, can we disable it in CodeClimate?

File: README.md
Comment (gilest) [reply README.md:12]: Ok sure thing 🤷

**Comment by jelhan** [comment]
> @gilest question about CodeClimate -- it looks like it's failing because it doesn't like our `.template-lintrc.js`, I'm guessing because its `ember-template-lint` implementation/integration is out-of-date? Now that template linting is part of the CI workflow, can we disable it in CodeClimate?

We discussed dropping CodeClimate in #429 as well because that would simplify the migration to GitHub Actions. I'm very much in favor of doing so. 👍

**Comment by jelhan** [comment]
I have mixed feelings about a `v4` branch. I see the benefits provided by keeping `master` backward compatible while working on next major in parallel. But it also has trade-offs. Especially increased maintenance costs as it would be a long living feature branch. It also comes with the risk of getting stale.

I would tend towards freezing v3 as soon as possible and starting work on v4 in `master`. We might want to release another minor version with Embroider support, fixed CI and migration from TravisCI to GitHub Actions. I think merge requests for all of that are pending. So as soon as possible might be very soon. 

If needed we could still release a hot fix from feature branch. But all further feature development would go in v4. This would also provide helpful pressure to actually ship it. E.g. to delay some nice to have features until v5 if needed.

**Comment by bendemboski** [comment]
> We discussed dropping CodeClimate in #429 as well because that would simplify the migration to GitHub Actions. I'm very much in favor of doing so. 👍

@jelhan I turned off template linting in the code climate config, and then it failed with too many errors to display. My guess was eslint, so I turned that off too, and now it's [failing](https://codeclimate.com/github/adopted-ember-addons/ember-file-upload/pull/430) with a small number of errors, but at least a few of them don't look like ones that I could have possibly introduced. Any thoughts? Or is it cool if I just remove CodeClimate entirely in this PR? Or would you prefer I open a separate PR for that so it (and the move to GitHub actions) could be applied to the `v3` line without all the breaking changes in this PR?

**Comment by gilest** [comment]
@bendemboski I've merged #439 which drops Code Climate entirely.

Give it a rebase and let's see how the CI runs

**Comment by bendemboski** [comment]
@gilest @jelhan I think we're good now. For some reason it's still complaining about `codeclimate`, but I'm guess that's just because GitHub checks are confused because this PR was first opened with codeclimate enabled, and now it's disabled. So I assume after merging, it will all be fine in `master`. If you want me to try closing and re-opening the PR or something, let me know.

I had to do some fiddling to get things working with CSP, but I think it turned out alright -- check out the last commit for details.

**Comment by gilest** [comment]
@bendemboski Thanks for your effort here 👏 

CI run looks great! ✅ 

Don't worry about the CC check. If it pops up in the next build/PR I will alert a maintainer who has permissions to remove the integration from GitHub.

### Release plan

I'm trying to resolve a timing/promise exception swallowing issue that is blocking the Github Actions PR currently (#429).

> There's a test failure in the mirage-handler test that checks for errors emitted from uploading an invalid audio file. Can sort-of reproduce by adding a small timeout or a debugger [here](https://github.com/adopted-ember-addons/ember-file-upload/blob/master/addon/mirage/utils.js#L118)

Do you have any idea why this might be occurring? It's failing the `ember-release` scenario of Travis builds also – but not in this PR – did you encounter it during the upgrade? Guess it's some incompatibility between package versions... 🤷 

Once that's been sorted I'd be comfortable with @jelhan cutting a final v3 release.

At that point we could merge this PR and begin releasing v4 alpha/betas.

**Comment by bendemboski** [comment]
I have not encountered any failures like that in the various builds I've run. It would not surprise me at all if among all of the [dependency updates](https://github.com/adopted-ember-addons/ember-file-upload/pull/430/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) (mostly from the blueprint changes from `ember-cli-update`), something was updated that fixed some intermittent issue. I also just looked at the `yarn.lock` changes and `ember-cli-mirage` and `miragejs` were each upgraded by a bugfix version or two -- the changelogs don't indicate anything that seems like an explicit fix, but `miragejs` has an "update dependencies" change...so there are plenty of possible explanations for why this branch maybe-fixes the intermittent issue. How's that for a statement of confidence?! 😆

**Comment by gilest** [comment]
Thanks – I tried with the minor bumps of mirage packages – no bueno 😅 . Will keep experimenting.

File: package.json
Comment (gilest) [inline package.json:51]: Could we trying dropping these in this PR. I think it should be safe. They were only added in #431

Comment (bendemboski) [reply package.json]: Yup, looks like it!

**Review by jelhan — APPROVED** [review]

**Review by gilest — APPROVED** [review]

**Comment by jelhan** [comment]
@bendemboski There are some merge conflicts. Do you have time to resolve them?

**Comment by bendemboski** [comment]
@jelhan done

**Comment by jelhan** [comment]
Thanks a lot!

**Pull request merged**
